### DISCUSSION TITLE
Extending Apollo Router

### DISCUSSION THREAD

**Discussion by Ronnie-personal** [apollographql/router#2581] [Q&A]
Can you please help me to understand the options for extending the Apollo router?
What are the pros and cons for of each option?

I came across this document 'https://www.apollographql.com/docs/router/configuration/external/' and I have also heard about 'Custom Middleware' and 'Custom Directives.'

I would appreciate it if you could provide me with more information.

Thank You!

**Comment by garypen** [comment]
There is a hierarchy of extensibility. You get more power as you move up the hierarchy, but each step requires more effort:

 1. It may be that what you want to do is built into the router and can be done by configuration: header propagation, telemetry, etc..
 2. If you want to do something slightly more complex, then `rhai` scripting is your next option: custom header concatentation, etc...
 3. If you have existing functionality or want to work in a different language, then external extensibility may be the right choice
 4. If you want to do "pretty much anything", then writing a rust plugin is your most powerful option.

The big advantage of choices: 1, 2 or 3 is that they don't require you to build your own router from source and you can just use our router binary.

I recommend reading all the documentation relating to:
 - [configuration](https://www.apollographql.com/docs/router/configuration/overview)
 - [customization](https://www.apollographql.com/docs/router/customizations/overview)
 - [external extensions](https://www.apollographql.com/docs/router/configuration/external/)

**Reply by Ronnie-personal** [reply]
@garypen  
Thank you for the prompt reply!
'External extensions' are experimental, so I guess they are not suitable for a production workload. 
Can you please clarify?

**Reply by bnjjj** [reply]
@Ronnie-personal Documentation about experimental feature is available [here](https://www.apollographql.com/docs/resources/product-launch-stages#experimental-features). TLDR it's more a disclaimer to say that we can change the configuration change and create breaking changes.
